File: src/output-budget.ts
import type { AppConfig } from "./types.js";

/**
 * Ceilings on what a single tool call may return.
 *
 * `exec_command` already had one, ported from Codex; the tools written before
 * it did not, and an uncapped `read_file` or `glob` is the fastest way to spend
 * a context window. These are defaults rather than the model's business: a
 * caller that forgets to paginate should still get a bounded answer.
 *
 * Every cut is announced in the tool's own text, with the argument that
 * continues from where it stopped. Silent truncation reads as "that was the
 * whole file", which is worse than no cap at all.
 */
export const DEFAULT_MAX_FILE_LINES = 1_000;
export const DEFAULT_MAX_FILE_BYTES = 131_072;
export const DEFAULT_MAX_ENTRIES = 500;
export const DEFAULT_MAX_TREE_NODES = 1_000;

export interface FileBudget {
  maxLines: number;
  maxBytes: number;
}

export function fileBudget(config: AppConfig): FileBudget {
  return {
    maxLines: config.output?.maxFileLines ?? DEFAULT_MAX_FILE_LINES,
    maxBytes: config.output?.maxFileBytes ?? DEFAULT_MAX_FILE_BYTES,
  };
}

export function entryBudget(config: AppConfig): number {
  return config.output?.maxEntries ?? DEFAULT_MAX_ENTRIES;
}

export function treeNodeBudget(config: AppConfig): number {
  return config.output?.maxTreeNodes ?? DEFAULT_MAX_TREE_NODES;
}

export interface FileWindow {
  /** The lines actually returned. */
  lines: string[];
  /** 0-based index of the first returned line. */
  start: number;
  /** Total lines in the file, before any windowing. */
  total: number;
  /** Human-readable note about what was cut, or null when nothing was. */
  notice: string | null;
}

/**
 * Slice a file to a window that fits both budgets.
 *
 * The byte cap is not redundant with the line cap: a minified bundle is often
 * one line several megabytes long, which a line cap alone would return in full.
 */
export function windowFileLines(
  lines: string[],
  offset: number,
  requestedLimit: number | undefined,
  budget: FileBudget,
): FileWindow {
  const total = lines.length;
  const start = Math.max(0, Math.min(Math.trunc(offset) || 0, total));
  const limit = Math.max(
    0,
    Math.min(requestedLimit ?? budget.maxLines, budget.maxLines),
  );

  const kept: string[] = [];
  let bytes = 0;
  let cutOnBytes = false;

  for (const line of lines.slice(start, start + limit)) {
    const size = Buffer.byteLength(line, "utf8") + 1;
    if (bytes + size > budget.maxBytes) {
      if (kept.length === 0) {
        // A single line larger than the whole budget: return a prefix of it so
        // the caller sees something rather than an empty window.
        kept.push(
          Buffer.from(line, "utf8").subarray(0, budget.maxBytes).toString("utf8"),
        );
      }
      cutOnBytes = true;
      break;
    }
    kept.push(line);
    bytes += size;
  }

  const end = start + kept.length;
  if (start === 0 && end === total && !cutOnBytes) {
    return { lines: kept, start, total, notice: null };
  }

  const reason = cutOnBytes ? ", cut at the byte budget" : "";
  const more = end < total ? ` — call again with offset=${end} for the rest` : "";
  return {
    lines: kept,
    start,
    total,
    notice: `(showing lines ${start + 1}-${end} of ${total}${reason}${more})`,
  };
}

/** Cut a list to `max`, reporting how many entries were dropped. */
export function limitList<T>(items: T[], max: number): { items: T[]; dropped: number } {
  if (max <= 0 || items.length <= max) return { items, dropped: 0 };
  return { items: items.slice(0, max), dropped: items.length - max };
}

File: src/output_budget.rs
//! Ceilings on what a single tool call may return. Ports `src/output-budget.ts`.
//!
//! Every cut is announced in the tool's own text with the argument that
//! continues from where it stopped; silent truncation reads as "that was the
//! whole file", which is worse than no cap at all.

use crate::types::AppConfig;

pub const DEFAULT_MAX_FILE_LINES: usize = 1_000;
pub const DEFAULT_MAX_FILE_BYTES: usize = 131_072;
pub const DEFAULT_MAX_ENTRIES: usize = 500;
pub const DEFAULT_MAX_TREE_NODES: usize = 1_000;

#[derive(Debug, Clone, Copy)]
pub struct FileBudget {
    pub max_lines: usize,
    pub max_bytes: usize,
}

pub fn file_budget(config: &AppConfig) -> FileBudget {
    FileBudget {
        max_lines: config
            .output
            .max_file_lines
            .unwrap_or(DEFAULT_MAX_FILE_LINES),
        max_bytes: config
            .output
            .max_file_bytes
            .unwrap_or(DEFAULT_MAX_FILE_BYTES),
    }
}

pub fn entry_budget(config: &AppConfig) -> usize {
    config.output.max_entries.unwrap_or(DEFAULT_MAX_ENTRIES)
}

pub fn tree_node_budget(config: &AppConfig) -> usize {
    config
        .output
        .max_tree_nodes
        .unwrap_or(DEFAULT_MAX_TREE_NODES)
}

#[derive(Debug, Clone)]
pub struct FileWindow {
    /// The lines actually returned.
    pub lines: Vec<String>,
    /// 0-based index of the first returned line.
    pub start: usize,
    /// Total lines in the file, before any windowing.
    pub total: usize,
    /// Human-readable note about what was cut, or `None` when nothing was.
    pub notice: Option<String>,
}

/// Take the first `max_bytes` bytes of `s`, decoding lossily so a multibyte
/// character straddling the cut becomes U+FFFD — matching Node's
/// `Buffer.subarray(0, n).toString("utf8")`.
fn byte_prefix(s: &str, max_bytes: usize) -> String {
    if s.len() <= max_bytes {
        return s.to_string();
    }
    String::from_utf8_lossy(&s.as_bytes()[..max_bytes]).into_owned()
}

/// Slice a file to a window that fits both budgets.
///
/// The byte cap is not redundant with the line cap: a minified bundle is often
/// one line several megabytes long, which a line cap alone would return in full.
pub fn window_file_lines(
    lines: &[&str],
    offset: usize,
    requested_limit: Option<usize>,
    budget: FileBudget,
) -> FileWindow {
    let total = lines.len();
    let start = offset.min(total);
    let limit = requested_limit
        .unwrap_or(budget.max_lines)
        .min(budget.max_lines);

    let mut kept: Vec<String> = Vec::new();
    let mut bytes = 0usize;
    let mut cut_on_bytes = false;

    let slice_end = start.saturating_add(limit).min(total);
    for line in &lines[start..slice_end] {
        let size = line.len() + 1;
        if bytes + size > budget.max_bytes {
            if kept.is_empty() {
                // A single line larger than the whole budget: return a prefix so
                // the caller sees something rather than an empty window.
                kept.push(byte_prefix(line, budget.max_bytes));
            }
            cut_on_bytes = true;
            break;
        }
        kept.push((*line).to_string());
        bytes += size;
    }

    let end = start + kept.len();
    if start == 0 && end == total && !cut_on_bytes {
        return FileWindow {
            lines: kept,
            start,
            total,
            notice: None,
        };
    }

    let reason = if cut_on_bytes {
        ", cut at the byte budget"
    } else {
        ""
    };
    let more = if end < total {
        format!(" — call again with offset={end} for the rest")
    } else {
        String::new()
    };
    let notice = format!(
        "(showing lines {}-{} of {}{}{})",
        start + 1,
        end,
        total,
        reason,
        more
    );
    FileWindow {
        lines: kept,
        start,
        total,
        notice: Some(notice),
    }
}

/// Cut a list to `max`, reporting how many entries were dropped.
pub fn limit_list<T>(mut items: Vec<T>, max: usize) -> (Vec<T>, usize) {
    if max == 0 || items.len() <= max {
        return (items, 0);
    }
    let dropped = items.len() - max;
    items.truncate(max);
    (items, dropped)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn budget(max_lines: usize, max_bytes: usize) -> FileBudget {
        FileBudget {
            max_lines,
            max_bytes,
        }
    }

    #[test]
    fn whole_small_file_has_no_notice() {
        let lines = vec!["a", "b", "c"];
        let w = window_file_lines(&lines, 0, None, budget(1000, 1000));
        assert_eq!(w.lines, vec!["a", "b", "c"]);
        assert!(w.notice.is_none());
    }

    #[test]
    fn line_cap_windows_and_notices() {
        let lines = vec!["a", "b", "c", "d"];
        let w = window_file_lines(&lines, 0, None, budget(2, 1000));
        assert_eq!(w.lines, vec!["a", "b"]);
        let n = w.notice.unwrap();
        assert!(n.contains("offset=2"), "{n}");
    }

    #[test]
    fn offset_windows() {
        let lines = vec!["a", "b", "c", "d"];
        let w = window_file_lines(&lines, 2, None, budget(1000, 1000));
        assert_eq!(w.lines, vec!["c", "d"]);
    }

    #[test]
    fn byte_cap_on_single_huge_line() {
        let huge = "x".repeat(100);
        let lines = vec![huge.as_str()];
        let w = window_file_lines(&lines, 0, None, budget(1000, 10));
        assert_eq!(w.lines[0].len(), 10);
        assert!(w.notice.unwrap().contains("byte budget"));
    }

    #[test]
    fn limit_list_reports_dropped() {
        let (items, dropped) = limit_list(vec![1, 2, 3, 4, 5], 3);
        assert_eq!(items, vec![1, 2, 3]);
        assert_eq!(dropped, 2);
    }
}
